### PR TITLE
VZ-10265 Check enabled istio deployments for readiness check

### PR DESCRIPTION
**Pull request by davisgomes**
**Changes**
- The is ready check uses label selectors to make sure the deployments for Istio are ready

**Testing**
- Deploy a cluster with egress gateway disabled and verify the readiness check works
- Deploy a cluster with all enabled and verify the readiness check works

### CODE REVIEW

File: platform-operator/controllers/verrazzano/component/istio/istio_component.go [hunk 1]
```diff
@@ -443,8 +455,70 @@ func (i istioComponent) IsReady(context spi.ComponentContext) bool {
...
+	})
+	err := ctx.Client().Get(context.TODO(), types.NamespacedName{Name: istioOperatorName, Namespace: IstioNamespace}, &istioOp)
+	if kerrs.IsNotFound(err) || meta.IsNoMatchError(err) {
+		return istioDeployments, nil
+	}
+	if err != nil {
+		return istioDeploy, ctx.Log().ErrorfNewErr("Failed to get the Istio Operator object from the cluster, %v", err)
+	}
+
+	if parseIstioOperatorJSONAsBool(istioOp.Object, "spec", "components", "egressGateways", 0, "enabled") {
```
[the remark this answers is not in the document]
Comment (davisgomes) [reply platform-operator/controllers/verrazzano/component/istio/istio_component.go]: Migrating to label selectors to help address some of these edge case concerns and to reduce complexity.

**Review by desagar — DISMISSED** [review]

File: platform-operator/controllers/verrazzano/component/istio/istio_component.go [hunk 2]
```diff
@@ -411,7 +396,7 @@ func (i istioComponent) Upgrade(context spi.ComponentContext) error {
 }
 
 func (i istioComponent) IsAvailable(ctx spi.ComponentContext) (reason string, available vzapi.ComponentAvailability) {
-	return (&ready.AvailabilityObjects{DeploymentNames: istioDeployments}).IsAvailable(ctx.Log(), ctx.Client())
+	return (&ready.AvailabilityObjects{DeploymentSelectors: []clipkg.ListOption{&istioLabelSelector}}).IsAvailable(ctx.Log(), ctx.Client())
```
Comment (pfmackin) [inline platform-operator/controllers/verrazzano/component/istio/istio_component.go:399]: what happens if this is called before any deployments are created, meaning the selector finds nothing.

Comment (davisgomes) [reply platform-operator/controllers/verrazzano/component/istio/istio_component.go:399]: The availability check handles an empty selector list like this:

File: platform-operator/controllers/verrazzano/component/istio/istio_component.go [hunk 3]
```diff
@@ -443,8 +428,9 @@ func (i istioComponent) IsReady(context spi.ComponentContext) bool {
 	return true
 }
 
-func areIstioDeploymentsReady(context spi.ComponentContext, prefix string) bool {
-	return ready.DeploymentsAreReady(context.Log(), context.Client(), istioDeployments, 1, prefix)
+// areIstioDeploymentsReady verifies that the enabled deployments in the Istio Operator are ready
```
Comment (pfmackin) [inline platform-operator/controllers/verrazzano/component/istio/istio_component.go:431]: This should be IstioOperator CR, not Istio Operator

File: platform-operator/controllers/verrazzano/component/istio/istio_component.go (hunk 2 above)
Comment (pfmackin) [reply platform-operator/controllers/verrazzano/component/istio/istio_component.go:399]: So this assumes that all deployments exist before any of them are ready.  That is probably the case but it is timing sensitive.

Comment (michael-cico) [reply platform-operator/controllers/verrazzano/component/istio/istio_component.go:399]: I would think that by the time `helm` finishes the `deployment` resources should exist right?

File: platform-operator/controllers/verrazzano/component/istio/istio_component.go (hunk 3 above)
Comment (davisgomes) [reply platform-operator/controllers/verrazzano/component/istio/istio_component.go]: Thanks fixed.

File: platform-operator/controllers/verrazzano/component/istio/istio_component.go (hunk 2 above)
Comment (davisgomes) [reply platform-operator/controllers/verrazzano/component/istio/istio_component.go:399]: Istioctl controls the Istio lifecycle, but I think your assumption is true at least a great amount of the time.

**Review by anders-swanson — APPROVED** [review]

**Review by pfmackin — APPROVED** [review]

File: platform-operator/controllers/verrazzano/component/istio/istio_component.go (hunk 2 above)
Comment (michael-cico) [reply platform-operator/controllers/verrazzano/component/istio/istio_component.go:399]: yeah... I forgot about Istio.  But, I think `istioctl` blocks until it finishes at least creating the resources (although we run it in the background), so by the time we get to the ready checks it should be there.

But yes, it's still a potential race condition.

**Pull request merged**